tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/source_code/source_code_component.ts

```diff
@@ -0,0 +1,139 @@
+ *
+ * TODO(cais): Add support for line decoration and symbol decoration.
+ *
+ * Unlike SourceFilesComponent, SourceCodeComponent handles only one file at a
+ * time.
+ */
+@Component({
+  selector: 'source-code-component',
+  templateUrl: './source_code_component.ng.html',
+  styleUrls: ['./source_code_component.css'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SourceCodeComponent {
+  // Lines of the source-code file, split at line breaks.
+  @Input()
+  lines: string[] | null = null; // TODO(cais): Add spinner for `null`.
+
+  // Line number to scroll to and highlight, 1-based.
+  @Input()
+  focusedLineno: number | null = null;
+
+  @ViewChild('codeViewerContainer', {static: true, read: ElementRef})
+  private readonly codeViewerContainer!: ElementRef<HTMLDivElement>;
+
+  private editor: any = null;
+
+  private decorations: string[] = [];
+
+  async ngOnChanges(changes: SimpleChanges): Promise<void> {
+    await loadMonaco();
+    const monaco = windowWithRequireAndMonaco.monaco;
+    let currentLines: string[] | null = this.lines;
+    if (changes.lines && changes.lines.currentValue !== null) {
+      currentLines = changes.lines.currentValue;
+      const value = changes.lines.currentValue.join('\n');
+      if (this.editor === null) {
+        this.editor = monaco.editor.create(
+          this.codeViewerContainer.nativeElement,
+          {
+            value,
+            language: DEFAULT_CODE_LANGUAGE,
+            readOnly: true,
+            fontSize: DEFAULT_CODE_FONT_SIZE,
+            minimap: {
+              enabled: true,
+            },
+          }
+        );
+      } else {
+        this.editor.setValue(value);
+      }
+    }
+    if (
+      changes.focusedLineno &&
+      changes.focusedLineno.currentValue &&
+      currentLines &&
+      this.editor !== null
+    ) {
+      this.editor.revealLineInCenter(
+        changes.focusedLineno.currentValue,
+        monaco.editor.ScrollType.Smooth
+      );
+      const lineLength =
+        currentLines[changes.focusedLineno.currentValue - 1].length;
+      this.decorations = this.editor.deltaDecorations(this.decorations, [
+        {
+          range: new monaco.Range(
+            changes.focusedLineno.currentValue,
+            1,
+            changes.focusedLineno.currentValue,
+            1
+          ),
+          options: {
+            isWholeLine: true,
+            linesDecorationsClassName: 'highlight-gutter',
+          },
+        },
+        {
+          range: new monaco.Range(
+            changes.focusedLineno.currentValue,
+            1,
+            changes.focusedLineno.currentValue,
+            lineLength + 1
+          ),
+          options: {
+            inlineClassName: 'highlight-line',
+          },
+        },
+      ]);
+    }
+  }
+
+  @HostListener('window:resize', ['$event'])
+  onResize(event: Event) {
+    if (this.editor !== null) {
+      this.editor.layout();
```

Comment:
Debounce this if possible. Resize can happen 60hz and we don't need to layout on every frame.